### ISSUE TITLE
Proposal: Allow `using` statement to structurally match `IDisposable`

### ISSUE THREAD

**Issue by agocke** [dotnet/csharplang#1623]
Currently the language states that any type used in an `using` statement must have an implicit conversion to `IDisposable`. This proposal is allow any type which structurally matches `IDisposable` (has a public void-returning instance method taking no arguments) to be used in a `using` statement.

The semantics of this construct are the same as a regular `using`, except that if the type does have an implicit conversion to `IDisposable`, there is a check to see if it structurally matches `IDisposable`. If so, instead of the `IDisposable.Dispose` call, there is a call to the structural `Dispose` method.

At the moment, extension methods are not considered, like `foreach` statements. If we were to add extension method support to `GetEnumerator` for `foreach`, I propose we also add it for `using`.

**Comment by agocke** [comment]
cc @gafter @jaredpar @fayrose

**Comment by CyrusNajmabadi** [comment]
> This proposal is allow any type which structurally matches IDisposable (has a public void-returning non-generic instance method taking no arguments) to be used in a using statement.

I would also like it if we could allow an appropriately matching extension method.  It seems reasonable to be able to embueue a type you don't own with RAII semantics if that makes sense for your domain.  

Just my 2c :)

**Comment by gafter** [comment]
I like it, but I would suggest that perhaps we should consider extension methods right from the start.

**Comment by CyrusNajmabadi** [comment]
Ha!  Beat you @gafter  :)

**Comment by gafter** [comment]
Not by much; we both submitted at 3:47.

**Comment by HaloFour** [comment]
Dupe of #93 😉

**Comment by HaloFour** [comment]
@agocke 

> The semantics of this construct are the same as a regular using, except that if the type does have an implicit conversion to `IDisposable`, there is a check to see if it structurally matches `IDisposable`. If so, instead of the `IDisposable.Dispose` call, there is a call to the structural `Dispose` method.

That should read:

"except that if the type **not** does have an implicit conversion to `IDisposable`"

Correct?  The compiler should only apply the structural typing if the type isn't `IDisposable` in order to not change the behavior of existing code, right?

**Comment by yaakov-h** [comment]
I think so.

Does "structually matches" also include extension methods, or do we need to explicitly mention that?

**Comment by Korporal** [comment]
Just my thoughts but the existing `dispose `pattern(s) are already pretty bewildering for many developers and this proposal effectively makes this even more abstruse. There are several scenarios to be considered when incorporating `dispose `support like "does the class inherit from a base class that already has dispose" or "do I actually 'own' the field that I want to ultimately dispose" and so on.

Perhaps the entire mechanics of `dispose `is ripe for a review and a simpler alternative mechanism be introduced...

**Comment by agocke** [comment]
@HaloFour Correct.

**Comment by andre-ss6** [comment]
Just a question: what would be the benefit of doing this if extensions methods were to be ignored? I'm curious.

**Comment by jnm2** [comment]
@andre-ss6 Besides consistency with the rest of the language, it would also be a stepping stone to enable `using await` to await a Dispose method that returns ValueTask or any other awaitable type.

**Comment by jaredpar** [comment]
@andre-ss6 

> Just a question: what would be the benefit of doing this if extensions methods were to be ignored? I'm curious.

One concrete benefit: allows `ref struct` to participate in `using` statements. There is no way for them to do this today as they can't implement interfaces by design.

**Comment by rvhuang** [comment]
Asking a stupid question: if this proposal comes true, what is the point of implementing `IDisposable` then?

**Comment by yaakov-h** [comment]
So that third-party code can safely dispose of your object... at least until shapes and `SDisposable` comes along.

**Comment by markrendle** [comment]
What if there are two extension classes in scope that both provide `public static void Dispose(this Thing thing)`? Currently when you get extension method clash, you get the `Call is ambiguous...` error; how would that be surfaced where the extension method invocation is masked by the `using` syntax?

**Comment by jnm2** [comment]
@markrendle The same thing that happens with conflicting GetAwaiter() extension methods and `await`, I suppose?

**Comment by HaloFour** [comment]
@markrendle 

> What if there are two extension classes in scope that both provide `public static void Dispose(this Thing thing)`? Currently when you get extension method clash, you get the `Call is ambiguous...` error; how would that be surfaced where the extension method invocation is masked by the `using` syntax?

This same situation can arise with `await` which may use extension methods to resolve `GetAwaiter` and the result is that same compilation error due to calls being ambiguous.

**Comment by jnm2** [comment]
In particular: [sharplab](https://sharplab.io/#v2:EYLgHgbALANALiAbgUwE4EsBmBPAPgAQAYACfARgDoAlAVwDs50BbZCgYQHsmAHdAGzQBlNInQBjZAGcA3AFgAUEVJkArHPkL8AZlIAmYmwUBvBcTOkd+ABykIxALIAKAJSnzJ+ea+kAnMTrIAO7EAGIcHC7qXgC+CrEaijqoyACGACYcdHzYxJJwqDRicKHhxEbE8ZqWZHb4+gCiYHDIdJLomZJkxgoAkNrKtRAAgoEp6M2oxADiyHAjYxOOcAAW6JIlHMSY4c7EALwAfLbsXNwCzWn4EBQzc6PjaJHmcQpVA3rEjc2t7a263fI+tVBvMHpNbqDFis1hstjt9kcriceOdkJdrhD7otnNJnvJokA)

**Comment by scalablecory** [comment]
> Correct? The compiler should only apply the structural typing if the type isn't IDisposable in order to not change the behavior of existing code, right?

I'd be curious to see what code would be affected by this that wasn't already buggy. Sounds like a big code smell.

**Comment by markrendle** [comment]
@jnm2 @HaloFour Right, yes, I forgot that await was structural already.

**Comment by GlassGrass** [comment]
Sorry for my annoying comment if it is already-known.

Though I'm sure this proposal is very good, I have to confirm backward compatibility.

For example, under current implementation, the following code:

```CSharp
using System;

public static class Program
{
    public static void Main()
    {
        using(var foo = new Foo())
        {
            Console.WriteLine("Some operation ...");
        }
    }
}

public class Foo : IDisposable
{
    public void Dispose()
    => Console.WriteLine("via implicit name resolution against 'Foo' type");

    void IDisposable.Dispose()
    => Console.WriteLine("via explicit name resolution against 'IDisposable' type");
}
```

calls calls `public void Dispose()`, but becomes to call `public void Dispose()` if the syntax suger like as `foreach` will be employed.

For maintaining backward compatibility, the naming resolution must be tried after the casting to `IDisposable` was failure.

**Comment by ufcpp** [comment]
@GlassGrass 
https://github.com/dotnet/csharplang/pull/1703/files#diff-21e04e0195b444dbbbe1d304197b460fR123

**Comment by jaredpar** [comment]
@GlassGrass that is explicitly covered in the proposal (@ufcpp has linked to it).

**Comment by GlassGrass** [comment]
@ufcpp @jaredpar 
I got it, thank you.

**Comment by LYP951018** [comment]
We just need Shapes.

**Comment by jnm2** [comment]
Now that `await using` has advanced, how soon until the structural matching allows `await using` on any type with a public or interface-implemented `DisposeAsync()` method with an awaitable return type with a void result?

**Comment by bartdesmet** [comment]
Also, given the following extension method

```csharp
public static ConfiguredCancelableAsyncEnumerable<T>.Enumerator ConfigureAwait<T>(this IAsyncEnumerator<T> enumerator, bool continueOnCapturedContext)
```

using the `ConfiguredCancelableAsyncEnumerable<T>.Enumerator` type that has shipped recently in .NET Core 3.0 Preview 2, to aid in writing code like

```csharp
await using (var e = source.GetAsyncEnumerator(cancellationToken).ConfigureAwait(false))
{
    // ...
}
```

when `await foreach` can't be used (e.g. because the logic requires direct control over `MoveNextAsync` calls, as is the case for many LINQ operators and aggregates), currently results in error

```
error CS8410: 'ConfiguredCancelableAsyncEnumerable<TSource>.Enumerator': type used in an async using statement must be implicitly convertible to 'System.IAsyncDisposable'
```

It'd be a pity to have to resort to `try...finally...` here in order to apply `ConfigureAwait` for purposes of "configured disposal".

For this to work as intended, `await using` would need to support binding to any `DisposeAsync` method, using an empty argument list (possibly filling in optional parameters or empty `params` arrays?), whose return type is awaitable. That way, we can achieve the intended effect of awaiting on the "configured" awaitable.

Also tagging @stephentoub given this discussion refers to the corefx types introduced for configuring async enumerables.

**Comment by stephentoub** [comment]
> await using would need to support binding to any DisposeAsync method

Isn't that the plan now / already implemented @jcouv?

**Comment by jcouv** [comment]
@bartdesmet Yes, we've arrived at that same conclusion (when @stephentoub implemented methods like `ConfigureAwait`).
Pattern-based disposal should work already (as of preview3). Both `using`, and `foreach` (as well as `await using` and `await foreach`) will bind a `.Dispose()` (or `.DisposeAsync()`) call with no arguments. This recognizes methods with `params` or optional parameters.
The only caveat is that extension methods will no contribute at the moment (ie. dev16 preview4). This may be revised by the time C# 8.0 ships.
Let me know if you run into any unexpected behavior.

For reference (if you want to look at some test cases), here's [the PR](https://github.com/dotnet/roslyn/pull/32731) for doing that in async constructs.

**Comment by bartdesmet** [comment]
Thanks, this works now as expected, and I've converted the LINQ code to use `await using` (with a "configured" enumerator) where necessary.

**Comment by IanKemp** [comment]
So, is this done and dusted? If so, can it be closed?

**Comment by agocke** [comment]
C# 8 hasn't shipped yet, so this should remain open until it has

**Comment by alrz** [comment]
>At the moment, extension methods are not considered, like foreach statements. If we were to add extension method support to GetEnumerator for foreach, I propose we also add it for using.

Is the "extension-based foreach" rejected indefinitely or it's just cut from this release?

**Comment by Bosch-Eli-Black** [comment]
Can this be closed, now that C# 8 has shipped?

**Comment by 333fred** [comment]
No, we keep issues open until a spec has been merged.

**Comment by andre-ss6** [comment]
But [hasn't it](https://github.com/dotnet/csharplang/pull/1703)?

**Comment by yaakov-h** [comment]
No. That adds a proposal (`proposals/` subfolder), but the spec itself (`spec/` subfolder) has not been updated to incorporate these (or _any_ recent) changes.
